man/man1/arraymoments.1
.TH ARRAYMOMENTS  1 "17 AUGUST 2006"  "" "TIPSY COMMANDS"
.SH NAME
arraymoments \- write the shape as computed in shells sorted by array values 
.SH SYNOPSIS
.B arraymoments
.I box-number
.I center-box-number
.I particle
.I binning
.I bin-number
.I filename
.I bin-count
.I increment
.I radius-bin-count
.SH DESCRIPTION
.B arraymoments
is a command that produces a file of name,
.I filename,
that contains various physical parameters that relate to the shape and
orientation of the density distribution for particles of type
.I particle
in box, 
.I box-number.
If the boxes
.I box-number
or
.I center-box-number
are not loaded an error is generated. An array must be loaded using the
.B readarray
command for particle grouping. 
The shapes are calculated by finding the principle moments of inertia 
for each bin of particles. Particles are grouped in bins based on
their array values in decreasing order.
The initial guess for the center of the density distribution is taken
to be the center of mass of box
.I center-box-number.
The 
.I binning
argument refers to how the particles are binned, either in one large
bin that expands its number of particles or in shells of equal
particle number.  There are
.I bin-count
particles in a shell and each successive bin
includes an additional number of particles equal to    
.I increment.
The number of particles used to calculate the radius of the bin is given by
.I radius-bin-count
and should generally be greater than 600 and must be less than or equal to 
.I bin-count.

The output file has the following columns: radius, b/a, c/a, phi, theta, psi,
center-x,center-y,center-z, S1, S2 and particlemax where b/a refers to the
ratio between the intermediate axis and the long axis and c/a refers to
the ratio between the short and long axes.  Phi, theta and psi refer to the
Euler angles that describe the orientation of the density distribution.
S1 and S2 are measures of prolateness and sphericity defined in
Babel and Starkman (Ap.J. 401:28, 1992).
Particlemax is the maximum rank of the particles in each bin. 
.SH OPTIONS
.SS Particle
.LP
The possible choices for
.I particle
are:
.RS
.TP 3
.B all
Use the all the particles in box
.I box-number
.TP 3
.B dark
Use only the dark matter particles in box
.I box-number
.TP 3
.B gas
Use only the gas particles in box
.I box-number
.TP 3
.B star
Use only the star particles in box
.I box-number
.RE
.SS Binning
.LP
The possible choices for
.I binning
are:
.RS
.TP 3
.B bin
Make 
.I bin-number
bins starting with 
.I bin-count 
particles and increasing the number of particles included by
.I increment
for each successive bin.
.TP 3
.B shell 
Make 
.I bin-number
shells containing 
.I bin-count
particles each and incrementing the maximum and minimum
particle ranking by 
.I increment
for each shell. 
.RE
.SH SEE ALSO
.BR profile (1),
.BR rotationcurve (1),
.SH BUGS
While any array could be read in for use with arraymoments,
the command was designed to be used with density arrays and so 
particles are selected from high to low array values. Thus when 
using array types that increase with radius the outer-most particles
will be binned first, and so the 
.I bin
option would not behave as desired. If necessary the array could
be raised to the -1 power before loading for sorting purposes.  


code/arraymoments.c
#include "defs.h"
#include "fdefs.h"

arraymoments(job)
    char job[MAXCOMM] ;
{
    char command[MAXCOMM] ;
    char particle_type[MAXCOMM] ;
    char bin_type[MAXCOMM] ;
    int box ;
    int center_box ;
    int number_bins ;
    Real center[MAXDIM] ;
    int i ;
    int bin_count ;

    int p_type;
    int *aindex;
    int aindex_size;
    int k, l, m;
    double p = log(3.0)/log(1.5);   
    double s1;
    double s3;
    int particlemax;
    int particlemin;
    int inc;
    int rbinsize;
    raverage = 0.0;
    particlemax = 0;
    particlemin = 0;
    rbinsize = 0; 

    if (sscanf(job,"%s %d %d %s %s %d %s %d %d %d",command,&box,&center_box,
	    particle_type,bin_type,&number_bins,
	    hardfile.name,&bin_count,&inc,&rbinsize) == 10) { 
	if (boxes_loaded[box] && boxes_loaded[center_box] &&
		number_bins <= 1000) {
	    if (strcmp(bin_type,"bin") == 0) {
	    }
	    else if (strcmp(bin_type,"shell") == 0) {
	    }
	    else {
		printf("<sorry, %s not a bin type, %s>\n",bin_type,title) ;
		return ;
	    }
	    if (strcmp(particle_type,"dark") != 0 && 
		    strcmp(particle_type,"gas") != 0 &&
		    strcmp(particle_type,"star") != 0 &&
		    strcmp(particle_type,"all") !=0) {
		printf("<sorry, %s is not a particle type, %s>\n",
			particle_type,title) ;
		return ;
	    }
	    if(bin_count < rbinsize) {
		printf("<sorry, need at least %d particles in a bin, %s>\n",rbinsize,title);
		return;
	    }

	    hardfile.ptr = fopen(hardfile.name,"w") ;

	    /* allocating and initializing index array */
	    if(array == NULL) { 
		printf("<sorry, no array read in yet, %s>\n",title);
		return;
	    }

	    if (strcmp(particle_type,"dark") == 0){
	        aindex_size = boxlist[box].ndark;
		p_type = DARK;
	    }
    	    else if (strcmp(particle_type,"star") == 0){
	        aindex_size = boxlist[box].nstar;
		p_type = STAR;
	    }
	    else if (strcmp(particle_type,"gas") == 0){
	        aindex_size = boxlist[box].ngas;
		p_type = GAS;
	    }
	    else{
	        aindex_size = boxlist[box].ndark + boxlist[box].nstar + boxlist[box].ngas;
		p_type = ALL;
	    }

	    if(inc*number_bins + bin_count - inc > aindex_size) {
		printf("<sorry, not enough particles for those bin settings, %s>\n",title);
		return; 
	    }

  	    aindex = (int *)malloc(aindex_size*sizeof(int));
	    if(aindex == NULL) {
		printf("<Error in creating index array, %s>\n",title);
		return;
	    }
	    for(k = 0; k < aindex_size; k++) {
	        aindex[k] = k;
	    }
	    sortarray(aindex, 0, aindex_size - 1, box, p_type);

	    /* for(m = lowershell ; m <= (lowershell + bin_count) && m < boxlist[box].ndark; m++) {
		mark_dark[boxlist[box].dpi[aindex[m]]]++;
    	    } */


	    for (i = 1 ;i <= number_bins ;i++) {
		if (strcmp(particle_type,"dark") == 0){
		    setvec(center,boxes[center_box].dark_com) ;
		}
		else if (strcmp(particle_type,"star") == 0){
		    setvec(center,boxes[center_box].star_com) ;
		}
		else if (strcmp(particle_type,"gas") == 0){
		    setvec(center,boxes[center_box].gas_com) ;
		}
		else{
		    setvec(center,boxes[center_box].total_com) ;
		}
		if (strcmp(bin_type,"bin") == 0) {
		    particlemin = 0;
		}
		else if (strcmp(bin_type,"shell") == 0) {
		    particlemin = inc*i - inc;	
		}
		particlemax = inc*i + bin_count - inc ;

		find_shape_array(particle_type,box,center,aindex,particlemax,particlemin,rbinsize) ;
		s1 = sin(PI/2.0*pow(1.0 - ca, p)) ;
		s3 = sin(PI/2.0*ba) ;
		fprintf(hardfile.ptr,"%g %g %g ",raverage,ba,ca) ;
		fprintf(hardfile.ptr,"%g %g %g ",phi,theta,psi) ;
		fprintf(hardfile.ptr,"%g %g %g ",center[0],
			center[1],center[2]) ;
		fprintf(hardfile.ptr,"%g %g %i\n",s1, s3, particlemax) ;
	    }
	    fclose(hardfile.ptr) ;
	    free(aindex);
	}
	else if (!boxes_loaded[box]) {
	    printf("<sorry, box %d is not loaded, %s>\n",box,title) ;
	}
	else if (!boxes_loaded[center_box]) {
	    printf("<sorry, box %d is not loaded, %s>\n",center_box,title) ;
	}
	else {
	    printf("<%d is too many bins, %s>\n",number_bins,title) ;
	}
    }
    else {
	input_error(command) ;
    }
}
